Improve websocket (re)connection

emschwartz opened the pull request:
* handle errors when sending subscription message
* make sure to (re)send the subscription message each time the plugin tries to connect
* reconnect faster when the websocket disconnects

resolves https://github.com/interledgerjs/ilp-plugin-bells/issues/113
replaces https://github.com/interledgerjs/ilp-plugin-bells/pull/117, for which the integration tests kept failing

emschwartz committed:
fix: handle error on ws.send

resolves https://github.com/interledgerjs/ilp-plugin-bells/issues/113

emschwartz committed:
fix: subscribe on (re)connect

the plugin should (re)send the subscribe notification every time it reconnects and the connect event should not be emitted until it gets confirmation that it is subscribed

emschwartz committed:
fix: reconnect websocket faster

connectors can lose money if the websocket is disconnected so the plugin should be much more aggressive about reconnecting it

sharafian reviewed (changes requested):
Current changes LGTM, but I think this error applies to the factory class too. If the factory's admin plugin disconnects, it should know to resubscribe all its accounts. Maybe add an `_reconnect` event that the factory can listen for and resubscribe on?

emschwartz committed:
fix: resubscribe factory on ws reconnect

based on comment from @sharafian https://github.com/interledgerjs/ilp-plugin-bells/pull/119\#pullrequestreview-31106428

emschwartz committed:
refactor(plugin): make connect function more readable

emschwartz committed:
fix(factory): .create resolves after subscribing

sharafian commented on src/lib/factory.js:
```diff
@@ -199,6 +207,7 @@ class PluginFactory extends EventEmitter2 {
     // delete all listeners to stop memory leaks
     this.plugins.get(username).removeAllListeners()
     this.plugins.delete(username)
+    // TODO should we resubscribe without that account included?
```
Yeah

sharafian commented on src/lib/factory.js:
```diff
@@ -178,16 +181,21 @@ class PluginFactory extends EventEmitter2 {
     plugin.disconnect = function () { return Promise.resolve(null) }
-    plugin.connect = function () { return Promise.resolve(null) }
+    plugin.connect = function () {
+      // TODO this should wait until the admin is subscribed to the account notifications
```
If `connect` is changed to only resolve once it's subscribed to accounts, we don't have to worry about this case

sharafian commented on src/lib/factory.js:
```diff
@@ -38,30 +38,30 @@ class PluginFactory extends EventEmitter2 {
+      this.adminPlugin.on('_rpc:notification', (notif) =>
+        co.wrap(this._routeNotification).call(this, notif))
+
+      this.adminPlugin.on('connect', () => {
```
This should be wrapped in a promise and yielded at some point, so that factory.connect won't resolve until after `this._subscribeAccounts()` is called in this callback.

emschwartz replied:
Changed it so that there is a `_subscribe` event that this will wait for when it connects initially

emschwartz committed:
fix(factory): remove resubscribes without account

other tests changed because this makes it so we will send subscription requests with no accounts included

emschwartz committed:
fix(factory): .connect resolves only after subscribing
